Yoast/wordpress-seo
# Add term hierarchy replacement variable on category page

Note: see the companion PR https://github.com/Yoast/wpseo-woocommerce/pull/546 that should be merged on `wpseo-woocommerce`.

## Context


* Allow users to use the taxonomy hierarchy as replacement variable on the taxonomy pages.

## Summary


This PR can be summarized in the following changelog entry:

* Adds the term ancestors hierarchy as a taxonomy replacement variable.

## Details
Given a structure, for example:
```
first level cat
    sub
        subsub
            mycat
```

the new replacement variable set on the `mycat` taxonomy will output ~only the parents e.g.:
`first level cat - sub - subsub`~ the whole tree

## Relevant technical choices:

- adds a new `term_hierarchy` replacement variable for _all_ the hierarchical taxonomies
- this includes the WordPress default post category taxonomy, the WooCommerce `product_cat`, and any other custom hierarchical taxonomy
- fixes a typo `term-in-custom-taxonomies` vs. `term-in-custom-taxonomy` please double check this, but I couldn't find any other reference to `term-in-custom-taxonomies`
- the term hierarchy is output only for _hierarchical_ taxonomy terms that do have ancestors
- however, the UI shows the replacement variable also for non-hierarchical custom taxonomies: this appears to depend on the way we add replacement variables for custom taxonomy where we're not distinguishing between hierarchical / non-hierarchical... will create a separate issue

## Test instructions

This PR can be tested by following these steps:

- install and activate WooCommerce, wpseo-woocommerce, and the Yoast Test Helper plugin
- switch wpseo-woocommerce to the `426-term-hierarchy-replacement-variable-on-category-page` branch
- in the WordPress admin menu, go to Tools > Yoast Test and enable post types & taxonomies.

**Product category:**
- create some WooCommerce categories with sub categories
- edit a sub-category
- in the sub-category snippet editor:
  - verify the `Term hierarchy` variable is displayed within the recommended variables visible by default when clicking "Insert snippet variable"
  - add the `Term hierarchy` variable to the SEO title or Meta description
  - see the snippet preview updates an shows the hierarchy
  - save and view the taxonomy page on the front end
  - inspect the page source and see the title or description contain the term hierarchy
  - don't forget to check og:title, twitter:title, og:description, twitter:description 
- go to SEO > Search appearance > Taxonomies > Product categories
- edit the default template and add the `Term hierarchy` variable to the SEO title or Meta description
- view a product category page where you didn't set custom variables so that it uses the default template
- inspect the source and check the term hierarchy is correctly output

**Repeat the steps above for:**
- post category
- book category
- movie category

**Test the separatpr:**
- go to SEO > Search Appearance > General and change the title separator
- verify the new separator is used in the term hierarchy replacement, both in the snippet preview and in the front end

As mentioned above, for now the variable is available also for **non-hierarchical custom** taxonomies. This appears to depend on the way we add variables to specific pages, see:

https://github.com/Yoast/wordpress-seo/blob/594b61ac87a3934965b1e0aae6b1696f60fd0722/admin/class-admin-editor-specific-replace-vars.php#L47-L50

Will create a specific issue for this. At the moment, the situation is as follows:
```
category                  hierarchical        term_hierarchy available        OK
post_tag                  non-hierarchical    term_hierarchy not available    OK
post_format               non-hierarchical    term_hierarchy not available    OK
product_cat               hierarchical        term_hierarchy available        OK
product_tag               non-hierarchical    term_hierarchy available        Not OK
product_shipping_class    non-hierarchical    term_hierarchy available        Not OK
book-category             hierarchical        term_hierarchy available        OK
book-genre                non-hierarchical    term_hierarchy available        Not OK
movie-category            hierarchical        term_hierarchy available        OK
movie-genre               non-hierarchical    term_hierarchy available        Not OK
```

## UI changes

## Documentation

## Quality assurance


Fixes https://github.com/Yoast/featurerequests/issues/426
